### ISSUE TITLE
Can google-ads be modified to work on Python 3.6?

### ISSUE THREAD

**Issue by infectious** [googleads/google-ads-python#319]
Hello,

We've written a data connector that uses version 3.3.0 of the google-ads package, talking to v2 of the Google Ads API. It runs as an ETL job on Google Composer (Airflow). Our issue is that because Google are sunsetting v2 of the Google Ads API on 21st Oct 2020, we need to change our code to talk to v3 (or above, but 3 will do). To do this, we need version 5.0.0 of the google-ads package. Now, from version 4.1.0 onwards, google-ads requires Python 3.7. This is a problem for us because Google Composer only runs on Python 3.6 and from what I've researched so far you can't change that.

So, my question is - can a version of google-ads be created that talks to the v3 (or higher) Google Ads API and works on Python 3.6.6?

Thanks,
Harry

**Comment by Just1B** [comment]
Hello Harry,

Why not use a [KubernetesPodOperator](https://cloud.google.com/composer/docs/how-to/using/using-kubernetes-pod-operator?hl=fr) and run your connector inside a container ?

**Comment by BenRKarl** [comment]
@infectious Unfortunately the library won't work with a version of Python that's lower than 3.7.0.

**Comment by infectious** [comment]
@BenRKarl Is that an un-changeable thing or is it feasible for you guys to create the equivalent 3.7 functionality under 3.6?

**Comment by infectious** [comment]
@Just1B Thanks for the suggestion. We've got some experience with PythonVirtualenvOperator so if we take that route, it'll probably be using that, unless the Kube operator has a compelling advantage.

**Comment by Just1B** [comment]
@infectious The KubernetesPodOperator is amazing 🤯

Possibilities are limitless but your team must be familiar with the docker / containers ecosystem and a little bit of kubernetes too.

**Comment by BenRKarl** [comment]
@infectious The primary reason this library isn't compatible with Python versions > 3.7.0 is that it uses a dynamic import mechanism that relies on a [module-level __getattr__](https://www.python.org/dev/peps/pep-0562/) method, (see [here](https://github.com/googleads/google-ads-python/blob/master/google/ads/google_ads/v5/__init__.py#L257)), which is only available in Python 3.7.0 and above.

It might possible to work around this by importing the generated proto modules directly, for example:

```py
# Recommended workflow for importing a Campaign instance.
from google.ads.google_ads.client import GoogleAdsClient
# Will trigger module-level __getattr__.
# Also instantiate the campaign object
campaign = GoogleAdsClient.get_type("Campaign")

# Alternative workflow
# Will not trigger module-level __getattr__
from google.ads.google_ads.v5.proto.resources.campaign_pb2 import Campaign
# You need to instantiate manually
campaign = Campaign()
```

It's not recommended that you import the proto modules directly because they're generated, not hand written, and it's possible that the structure of the library will change, breaking your import paths. Using `client.get_service` and `client.get_type` is a safer and more maintainable interface.

So in short - it is not changeable for the library to go back and support 3.6.6, but it's possible with some (not-recommended) effort and risk to make it work.

**Comment by infectious** [comment]
Thanks @BenRKarl
